symbolic-debuginfo/src/wasm/parser.rs

```diff
@@ -0,0 +1,252 @@
+
+                for import in isr {
+                    let import = import?;
+                    if let ImportSectionEntryType::Function(id) = import.ty {
+                        if !func_sigs
+                            .as_bitslice()
+                            .get(id as usize)
+                            .as_deref()
+                            .unwrap_or(&false)
+                        {
+                            return Err(WasmError::UnknownFunctionType);
+                        }
+
+                        num_imported_funcs += 1;
+                    }
+                }
+            }
+            // The function section declares all of the local functions present in the module
+            Payload::FunctionSection(fsr) => {
+                validator.function_section(&fsr)?;
+
+                if fsr.get_count() > 0 {
+                    kind = ObjectKind::Library;
+                }
+
+                funcs.resize(
+                    fsr.get_count() as usize,
+                    Symbol {
+                        name: None,
+                        address: 0,
+                        size: 0,
+                    },
+                );
+
+                // We actually don't care about the type signature of the function, other than that
+                // they exist
+                for id in fsr {
+                    if !func_sigs
+                        .as_bitslice()
+                        .get(id? as usize)
+                        .as_deref()
+                        .unwrap_or(&false)
+                    {
+                        return Err(WasmError::UnknownFunctionType);
+                    }
+                }
+            }
+            // The code section contains the actual function bodies, this payload is emitted at
+            // the beginning of the section. This one is important as the code section offset is
+            // used to calculate relative addresses in a `DwarfDebugSession`
+            Payload::CodeSectionStart { range, count, .. } => {
+                code_offset = range.start as u64;
+                validator.code_section_start(count, &range)?;
+            }
+            // We get one of these for each local function body
+            Payload::CodeSectionEntry(body) => {
+                let validator = validator.code_section_entry()?;
+
+                let (address, size) = get_function_info(body, validator)?;
+
+                funcs[body_index].address = address;
+                funcs[body_index].size = size;
+
+                // Though we have an accurate? size of the function body, the old method of symbol
+                // iterating with walrus extends the size of each body to be contiguous with the
+                // next function, so we do the same, other than the final function
+                if body_index > 0 {
+                    funcs[body_index - 1].size = address - funcs[body_index - 1].address;
+                }
+
+                body_index += 1;
```

Comment:
so if I understand correctly, the `FunctionSection` gives you the total number of sections? And then each `CodeSectionEntry` fills in the details?

I would rather just `push` here, and use `if let Some() = funcs.last()` for the check above.

Or is the problem rather that you could potentially have a `CustomSection("name")` that wants to fill out the names before we actually get to the `CodeSectionEntry`?

This feels a bit brittle, and you can easily panic if you create a wasm file without a `FunctionSection` (or one with a 0 count), but with a bunch of `CodeSectionEntry`s.

Reply:
Yes, the function section gives the total number of locally defined functions in the module, I believe it always has to come before the actual code section where the function bodies reside, but I can change it so it won't panic if that is the case. I also don't think the `name` custom section can come before the code section, but I'm not certain. If there was a case where there is a code section entry without a valid function section before it with a matching count of local functions, the wasm validation will fail before we get to the indexing, but I can change that just in case.